internal/controller/podschedulingpolicy_controller.go

```diff
@@ -0,0 +1,211 @@
+		}
+	}()
+
+	if err = r.ensureFinalizers(ctx, len(dbList.Items) > 0, psp); err != nil {
+		logger.Error(err, fmt.Sprintf("failed to update finalizers for pod scheduling policy='%s'", psp.GetName()))
+		return ctrl.Result{}, errors.Join(err, fmt.Errorf("failed to update finalizers for pod scheduling policy='%s': %w", psp.GetName(), err))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PodSchedulingPolicyReconciler) ensureFinalizers(ctx context.Context, used bool, psp *everestv1alpha1.PodSchedulingPolicy) error {
+	currentFinalizers := psp.GetFinalizers()
+	var desiredFinalizers []string
+	if used {
+		desiredFinalizers = append(currentFinalizers, everestv1alpha1.UsedResourceFinalizer)
+		// remove duplicates
+		desiredFinalizers = slices.Compact(desiredFinalizers)
+	} else {
+		// Remove the finalizer
+		desiredFinalizers = append([]string{}, psp.GetFinalizers()...)
+		desiredFinalizers = slices.DeleteFunc(desiredFinalizers, func(f string) bool {
+			return f == everestv1alpha1.UsedResourceFinalizer
+		})
+	}
+
+	slices.Sort(currentFinalizers)
+	slices.Sort(desiredFinalizers)
+
+	if !slices.Equal(currentFinalizers, desiredFinalizers) {
+		psp.SetFinalizers(desiredFinalizers)
+		if err := r.Update(ctx, psp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodSchedulingPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Predicate to trigger reconciliation only on .spec.podSchedulingPolicyName changes in the DatabaseCluster resource.
+	dbClusterEventsPredicate := predicate.Funcs{
+		// Allow create events only if the .spec.podSchedulingPolicyName is set
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Object.(*everestv1alpha1.DatabaseCluster).Spec.PodSchedulingPolicyName != ""
+		},
+
+		// Only allow updates when the .spec.podSchedulingPolicyName of the DatabaseCluster resource changes
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldObj := e.ObjectOld.(*everestv1alpha1.DatabaseCluster)
+			newObj := e.ObjectNew.(*everestv1alpha1.DatabaseCluster)
+
+			// Trigger reconciliation only if the .spec.podSchedulingPolicyName field has changed
+			return oldObj.Spec.PodSchedulingPolicyName != newObj.Spec.PodSchedulingPolicyName
+		},
+
+		// Allow delete events only if the .spec.podSchedulingPolicyName is set
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Object.(*everestv1alpha1.DatabaseCluster).Spec.PodSchedulingPolicyName != ""
+		},
+
+		// Nothing to process on generic events
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	pspEventsPredicate := predicate.Funcs{
+		// Nothing to process on create events
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+
+		// Allow update events only if the policy is in the system namespace.
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.ObjectOld.GetNamespace() == common.SystemNamespace
```

Comment:
Same as https://github.com/percona/everest-operator/pull/737/files#r2076015522 , I'd suggest using the value of `SystemNamespace` read from `cmd/main.go` rather than hard-coding